akka-stream/src/main/scala/akka/stream/impl/io/ByteStringParser.scala

```diff
@@ -22,54 +22,84 @@ private[akka] abstract class ByteStringParser[T] extends GraphStage[FlowShape[By
     private var current: ParseStep[T] = FinishedParser
     private var acceptUpstreamFinish: Boolean = true
+    private var untilCompact = CompactionThreshold
 
     final protected def startWith(step: ParseStep[T]): Unit = current = step
 
-    @tailrec private def doParse(): Unit =
+    /**
+     * doParse() is the main driver for the parser. It can be called from onPush, onPull and onUpstreamFinish.
+     * The general logic is that invocation of this method either results in an emitted parsed element, or an indication
+     * that there is more data needed.
+     *
+     * On completion there are various cases:
+     *  - buffer is empty: parser accepts completion or fails.
+     *  - buffer is non-empty, we wait for a pull. This might result in a few more onPull-push cycles, served from the
+     *    buffer. This can lead to two conditions:
+     *     - drained, empty buffer. This is either accepted completion (acceptUpstreamFinish) or a truncation.
+     *     - parser demands more data than in buffer. This is always a truncation.
+     */
+    private def doParse(): Unit = {
       if (buffer.nonEmpty) {
         val reader = new ByteReader(buffer)
-        val cont = try {
+        try {
           val parseResult = current.parse(reader)
           acceptUpstreamFinish = parseResult.acceptUpstreamFinish
-          parseResult.result.map(emit(objOut, _))
+          parseResult.result.foreach(push(objOut, _))
+
           if (parseResult.nextStep == FinishedParser) {
             completeStage()
-            false
           } else {
             buffer = reader.remainingData
             current = parseResult.nextStep
-            true
           }
         } catch {
           case NeedMoreData ⇒
             acceptUpstreamFinish = false
             if (current.canWorkWithPartialData) buffer = reader.remainingData
-            pull(bytesIn)
-            false
+
+            // Not enough data in buffer and upstream is closed
+            if (isClosed(bytesIn)) current.onTruncation()
+            else pull(bytesIn)
         }
-        if (cont) doParse()
-      } else pull(bytesIn)
+      } else {
+        if (isClosed(bytesIn)) {
+          // Buffer is empty and upstream is done. If the current phase accepts completion we are done,
+          // otherwise report truncation.
+          if (acceptUpstreamFinish) completeStage()
+          else current.onTruncation()
+        } else pull(bytesIn)
+      }
+    }
+
+    // Completion is handled by doParse as the buffer either gets empty after this call, or the parser requests
+    // data that we can no longer provide (truncation).
+    override def onPull(): Unit = doParse()
 
     def onPush(): Unit = {
-      pullOnParserRequest = false
+      // Buffer management before we call doParse():
+      //  - append new bytes
+      //  - compact buffer if necessary
       buffer ++= grab(bytesIn)
+      untilCompact -= 1
+      if (untilCompact == 0) {
```

Comment:
Basically the threshold is a pruning logic that amortizes copy cost but keeps ByteString nesting levels at a bounded level.